internal/k8s/controllers/gateway_controller.go

```diff
@@ -93,19 +94,62 @@ func (r *GatewayReconciler) SetupWithManager(mgr ctrl.Manager) error {
 			handler.EnqueueRequestsFromMapFunc(podToGatewayRequest),
 			builder.WithPredicates(predicate),
 		).
+		Watches(
+			&source.Kind{Type: &gateway.ReferencePolicy{}},
+			handler.EnqueueRequestsFromMapFunc(r.referencePolicyToGatewayRequests),
+		).
 		Complete(gatewayclient.NewRequeueingMiddleware(r.Log, r))
 }
 
 func podToGatewayRequest(object client.Object) []reconcile.Request {
-	gateway, managed := utils.IsManagedGateway(object.GetLabels())
+	gw, managed := utils.IsManagedGateway(object.GetLabels())
 
 	if managed {
 		return []reconcile.Request{
 			{NamespacedName: types.NamespacedName{
-				Name:      gateway,
+				Name:      gw,
 				Namespace: object.GetNamespace(),
 			}},
 		}
 	}
 	return nil
 }
+
+func (r *GatewayReconciler) referencePolicyToGatewayRequests(object client.Object) []reconcile.Request {
+	refPolicy := object.(*gateway.ReferencePolicy)
+
+	gateways := r.getGatewaysAffectedByReferencePolicy(refPolicy)
+
+	requests := make([]reconcile.Request, 0, len(gateways))
+
+	for _, gw := range gateways {
+		requests = append(requests, reconcile.Request{
+			NamespacedName: types.NamespacedName{
+				Name:      gw.Name,
+				Namespace: gw.Namespace,
+			},
+		})
+	}
+
+	return requests
+}
+
+// getGatewaysAffectedByReferencePolicy retrieves all Gateways potentially impacted by the ReferencePolicy
+// modification. Currently, this is unfiltered and so returns all Gateways in the namespace referenced by
+// the ReferencePolicy.
+func (r *GatewayReconciler) getGatewaysAffectedByReferencePolicy(refPolicy *gateway.ReferencePolicy) []gateway.Gateway {
+	var matches []gateway.Gateway
+
+	for _, from := range refPolicy.Spec.From {
+		// TODO: search by from.Group and from.Kind instead of assuming this ReferencePolicy references a Gateway
```

Comment:
Could we open an issue tracking this TODO now that we have three different controllers implementing similar logic? Could be a good first issue for a potential contributor.

Reply:
I might just contribute the change - I think it amounts to just early outing before querying all Gateways, but I didn't want to throw that into this PR and complicate testing and whatnot. Thoughts?

```go
if from.Group != "" || from.Kind != "Secret" {
    continue
}
```